File: API/1980_Adaptive_Market_Level/PY/adaptive_market_level_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")
clr.AddReference("StockSharp.Algo.Indicators")
clr.AddReference("StockSharp.Algo.Strategies")

from System import TimeSpan, Math
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import SimpleMovingAverage
from StockSharp.Algo.Strategies import Strategy


class adaptive_market_level_strategy(Strategy):

    def __init__(self):
        super(adaptive_market_level_strategy, self).__init__()

        self._period = self.Param("Period", 14) \
            .SetDisplay("Period", "SMA period", "Indicator")
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromHours(4))) \
            .SetDisplay("Candle Type", "Type of candles", "General")

        self._prev_sma = 0.0
        self._prev_prev_sma = 0.0
        self._count = 0

    @property
    def Period(self):
        return self._period.Value

    @Period.setter
    def Period(self, value):
        self._period.Value = value

    @property
    def CandleType(self):
        return self._candle_type.Value

    @CandleType.setter
    def CandleType(self, value):
        self._candle_type.Value = value

    def OnStarted2(self, time):
        super(adaptive_market_level_strategy, self).OnStarted2(time)

        sma = SimpleMovingAverage()
        sma.Length = self.Period

        self.SubscribeCandles(self.CandleType) \
            .Bind(sma, self.ProcessCandle) \
            .Start()

    def ProcessCandle(self, candle, sma_value):
        if candle.State != CandleStates.Finished:
            return

        sma_val = float(sma_value)
        self._count += 1

        if self._count < 3:
            self._prev_prev_sma = self._prev_sma
            self._prev_sma = sma_val
            return

        turn_up = self._prev_sma < self._prev_prev_sma and sma_val > self._prev_sma
        turn_down = self._prev_sma > self._prev_prev_sma and sma_val < self._prev_sma

        if turn_up and self.Position <= 0:
            if self.Position < 0:
                self.BuyMarket()
            self.BuyMarket()
        elif turn_down and self.Position >= 0:
            if self.Position > 0:
                self.SellMarket()
            self.SellMarket()

        self._prev_prev_sma = self._prev_sma
        self._prev_sma = sma_val

    def OnReseted(self):
        super(adaptive_market_level_strategy, self).OnReseted()
        self._prev_sma = 0.0
        self._prev_prev_sma = 0.0
        self._count = 0

    def CreateClone(self):
        return adaptive_market_level_strategy()

File: API/1901-2000/1980_Adaptive_Market_Level/CS/AdaptiveMarketLevelStrategy.cs
using System;
using System.Collections.Generic;

using Ecng.Common;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// Adaptive market level strategy using SMA slope direction changes.
/// </summary>
public class AdaptiveMarketLevelStrategy : Strategy
{
	private readonly StrategyParam<int> _period;
	private readonly StrategyParam<DataType> _candleType;

	private decimal _prevSma;
	private decimal _prevPrevSma;
	private int _count;

	public int Period { get => _period.Value; set => _period.Value = value; }
	public DataType CandleType { get => _candleType.Value; set => _candleType.Value = value; }

	public AdaptiveMarketLevelStrategy()
	{
		_period = Param(nameof(Period), 14)
			.SetGreaterThanZero()
			.SetDisplay("Period", "SMA period", "Indicator");

		_candleType = Param(nameof(CandleType), TimeSpan.FromHours(4).TimeFrame())
			.SetDisplay("Candle Type", "Type of candles", "General");
	}

	/// <inheritdoc />
	public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
		=> [(Security, CandleType)];

	/// <inheritdoc />
	protected override void OnReseted()
	{
		base.OnReseted();
		_prevSma = 0;
		_prevPrevSma = 0;
		_count = 0;
	}

	/// <inheritdoc />
	protected override void OnStarted2(DateTime time)
	{
		base.OnStarted2(time);

		var sma = new SimpleMovingAverage { Length = Period };

		SubscribeCandles(CandleType)
			.Bind(sma, ProcessCandle)
			.Start();
	}

	private void ProcessCandle(ICandleMessage candle, decimal smaValue)
	{
		if (candle.State != CandleStates.Finished)
			return;

		_count++;

		if (_count < 3)
		{
			_prevPrevSma = _prevSma;
			_prevSma = smaValue;
			return;
		}

		var turnUp = _prevSma < _prevPrevSma && smaValue > _prevSma;
		var turnDown = _prevSma > _prevPrevSma && smaValue < _prevSma;

		if (turnUp && Position <= 0)
		{
			if (Position < 0) BuyMarket();
			BuyMarket();
		}
		else if (turnDown && Position >= 0)
		{
			if (Position > 0) SellMarket();
			SellMarket();
		}

		_prevPrevSma = _prevSma;
		_prevSma = smaValue;
	}
}
